cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1616,52 +1657,70 @@ void reader::impl::allocate_columns(size_t skip_rows, size_t num_rows, bool uses
 
   // compute output column sizes by examining the pages of the -input- columns
   if (has_lists) {
-    auto& page_keys  = _chunk_itm_data.page_keys;
     auto& page_index = _chunk_itm_data.page_index;
+
+    hostdevice_vector<input_col_info> input_cols{_input_columns.size(), _stream};
+    size_t max_depth = 0;
+    for (size_t i = 0; i < _input_columns.size(); i++) {
+      auto depth                  = _input_columns[i].nesting_depth();
+      max_depth                   = depth > max_depth ? depth : max_depth;
+      input_cols[i].nesting_depth = depth;
+      input_cols[i].schema_idx    = _input_columns[i].schema_idx;
+    }
+    input_cols.host_to_device(_stream);
```

Comment:
Couple of suggestions here: if the data is only copied once, we should not use a `hostdevice_vector`; the loop can be replaced with `std::transform` + `std::max_element`
Something like

```
    std::vector<input_col_info> h_cols_info;
    h_cols_info.reserve(_input_columns.size());
    std::transform(_input_columns.cbegin(), _input_columns.cend(), 
    std::back_inserter(h_cols_info), 
        [](auto & col){return {depth,input_columns[i].schema_idx}; });
    auto const max_depth = std::max_element(h_cols_info.cbegin(), h_cols_info.cend(), 
        [](auto& l, auto& r){return l.nesting_depth < r.nesting_depth;});
    auto const d_cols_info = make_device_uvector_async(h_cols_info, stream, rmm::mr::get_current_device_resource());
```